contracts/router/quoter/DefaultPoolCalc.sol

```diff
@@ -0,0 +1,187 @@
+    // blob/5d538c47115c29990dea5aa8af679bd024c82e35/contracts/SwapUtilsV2.sol#L717
+
+    /// @inheritdoc IDefaultPoolCalc
+    function calculateAddLiquidity(address pool, uint256[] memory amounts) external view returns (uint256 amountOut) {
+        // Get the provided number of tokens. We will later verify that it matches actual number of tokens in the pool.
+        uint256 numTokens = amounts.length;
+        // (1) Verify that `pool.tokens.length <= numTokens`
+        _verifyTokensLengthCeiling(pool, numTokens);
+        (, , , , uint256 swapFee, , address lpToken) = IDefaultExtendedPool(pool).swapStorage();
+        // current state
+        ManageLiquidityInfo memory v = ManageLiquidityInfo({
+            d0: 0,
+            d1: 0,
+            preciseA: IDefaultExtendedPool(pool).getAPrecise(),
+            totalSupply: IERC20Metadata(lpToken).totalSupply(),
+            balances: new uint256[](numTokens),
+            multipliers: new uint256[](numTokens)
+        });
+        uint256[] memory newBalances = new uint256[](numTokens);
+        // (2) If `pool.tokens.length < numTokens`, the loop will revert when reading token at index `numTokens - 1`.
+        for (uint256 i = 0; i < numTokens; ++i) {
+            address token = IDefaultExtendedPool(pool).getToken(uint8(i));
+            v.balances[i] = IDefaultExtendedPool(pool).getTokenBalance(uint8(i));
+            newBalances[i] = v.balances[i] + amounts[i];
+            v.multipliers[i] = 10**(18 - IERC20Metadata(token).decimals());
+        }
+        // (1) + (2) implies that `pool.tokens.length == numTokens`
+        if (v.totalSupply != 0) {
+            v.d0 = _getD(_xp(v.balances, v.multipliers), v.preciseA);
+        } else {
+            for (uint256 i = 0; i < numTokens; ++i) {
+                require(amounts[i] > 0, "Must supply all tokens in pool");
+            }
+        }
+
+        // invariant after change
+        v.d1 = _getD(_xp(newBalances, v.multipliers), v.preciseA);
+        require(v.d1 > v.d0, "D should increase");
+
+        if (v.totalSupply == 0) {
+            return v.d1;
+        } else {
+            uint256 feePerToken = _feePerToken(swapFee, numTokens);
+            for (uint256 i = 0; i < numTokens; ++i) {
+                uint256 idealBalance = (v.d1 * v.balances[i]) / v.d0;
+                uint256 fees = (feePerToken * _difference(idealBalance, newBalances[i])) / FEE_DENOMINATOR;
+                newBalances[i] -= fees;
+            }
+            v.d1 = _getD(_xp(newBalances, v.multipliers), v.preciseA);
+            return ((v.d1 - v.d0) * v.totalSupply) / v.d0;
+        }
+    }
+
+    /// @dev Verifies that `pool.tokens.length <= numTokens`. Reverts if `pool.tokens.length > numTokens`.
+    function _verifyTokensLengthCeiling(address pool, uint256 numTokens) internal view {
+        // We will attempt to read the token at index `numTokens`.
+        try IDefaultExtendedPool(pool).getToken(uint8(numTokens)) returns (address) {
```

Comment:
Saddle checks `numTokens` is exactly equal to pool `pool.tokens.length`. Are there any issues that come with having instead `pool.tokens.length <= numTokens`?